### PR TITLE
MEMTRACK: added memory limit

### PR DESCRIPTION
**Pull request by hoopoepg**
- added memory limit to memtrack infra
  to allow terminate process on too
  high memory usage

### CODE REVIEW

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

File: src/ucs/config/global_opts.c [hunk 1]
```diff
@@ -41,6 +41,7 @@ ucs_global_opts_t ucs_global_opts = {
     .stats_dest            = "",
     .tuning_path           = "",
     .memtrack_dest         = "",
+    .memtrack_limit        = UINT64_MAX,
```
Comment (yosefe) [inline src/ucs/config/global_opts.c:44]: UCS_MEMUNITS_INF

File: src/ucs/debug/memtrack.c [hunk 2]
```diff
@@ -141,6 +206,12 @@ void ucs_memtrack_allocated(void *ptr, size_t size, const char *name)
     /* update specific and global entries */
     ucs_memtrack_entry_update(entry, size);
     ucs_memtrack_entry_update(&ucs_memtrack_context.total, size);
+    if (ucs_memtrack_context.total.size >= ucs_global_opts.memtrack_limit) {
+        ucs_memtrack_generate_report();
+        ucs_memunits_to_str(ucs_global_opts.memtrack_limit, limit_str,
+                            sizeof(limit_str));
+        ucs_fatal("Reached memtrack memory limit %s", limit_str);
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:213]: small case

**Review by yosefe — COMMENTED** [review]
can we add some CI for it?

**Comment by yosefe** [comment]
@hoopoepg can we add some CI for it?

**Comment by hoopoepg** [comment]
is there any way to hook ups_fatal?

**Comment by yosefe** [comment]
i'd suggest a small app which allocates too much mem, and examine its output in CI script

**Review by dmitrygx — APPROVED** [review]

File: contrib/test_jenkins.sh [hunk 3]
```diff
@@ -941,6 +941,16 @@ test_memtrack() {
 	UCX_MEMTRACK_DEST=stdout ./test/gtest/gtest --gtest_filter=test_memtrack.sanity
 }
 
+test_memtrack_limit() {
+	../contrib/configure-devel --prefix=$ucx_inst
+	make_clean
+	$MAKEP
+
+	echo "==== Running memtrack limit test ===="
+	UCX_MEMTRACK_DEST=stdout UCX_HANDLE_ERRORS=none UCX_MEMTRACK_LIMIT=512MB ./test/apps/test_memtrack_limit | grep -q 'SUCCESS'
```
Comment (yosefe) [inline contrib/test_jenkins.sh:950]: `... test_memtrack_limit |& grep -C 100 'SUCCESS'`
- use |& to redirect all output streams
- print surrounding lines to show more info in test log
(same for 'reached' grep)

File: src/ucs/debug/memtrack.c [hunk 4]
```diff
@@ -210,6 +210,10 @@ void ucs_memtrack_allocated(void *ptr, size_t size, const char *name)
         ucs_memtrack_generate_report();
         ucs_memunits_to_str(ucs_global_opts.memtrack_limit, limit_str,
                             sizeof(limit_str));
+        /* disable memtrack to prevent hang */
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:213]: you mean that failure handler may call memtrack recursively?

File: test/apps/test_memtrack_limit.c [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+{
+    ucp_params_t params = {
+        .field_mask = UCP_PARAM_FIELD_FEATURES,
+        .features   = UCP_FEATURE_TAG
+    };
+    ucs_status_t ctx_status;
+    ucp_context_h context;
+    void *ptr;
+
+    ctx_status = ucp_init(&params, NULL, &context);
```
Comment (yosefe) [inline test/apps/test_memtrack_limit.c:25]: why do we need to create ucp context?

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

**Comment by yosefe** [comment]
> @yosefe ok to squash?

Yes

**Comment by hoopoepg** [comment]
@yosefe back it port into integration branch?

**Comment by yosefe** [comment]
yes

**Pull request merged**